src/VisualStudio/Core/Def/Implementation/LanguageClient/AbstractLanguageServerClient.cs

```diff
@@ -0,0 +1,171 @@
+        public event AsyncEventHandler<EventArgs> StopAsync;
+#pragma warning restore CS0067 // event never used
+
+        public AbstractLanguageServerClient(
+            Workspace workspace,
+            LanguageServerClientEventListener eventListener,
+            IAsynchronousOperationListenerProvider listenerProvider,
+            string languageServerName,
+            string clientName)
+        {
+            _workspace = workspace;
+            _eventListener = eventListener;
+            _asyncListener = listenerProvider.GetListener(FeatureAttribute.FindReferences);
+
+            _clientName = clientName;
+            _languageServerName = languageServerName;
+        }
+
+        public async Task<Connection> ActivateAsync(CancellationToken cancellationToken)
+        {
+            var client = await _workspace.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                // there is no OOP. either user turned it off, or process got killed.
+                return null;
+            }
+
+            var hostGroup = new HostGroup(client.ClientId);
+            var hubClient = new HubClient(_clientName);
+
+            var stream = await ServiceHubRemoteHostClient.Connections.RequestServiceAsync(
+                _workspace,
+                hubClient,
+                _languageServerName,
+                hostGroup,
+                TimeSpan.FromMinutes(60),
+                cancellationToken).ConfigureAwait(false);
+
+            return new Connection(stream, stream);
+        }
+
+        /// <summary>
+        /// Signals that the extension has been loaded.  The server can be started immediately, or wait for user action to start.  
+        /// To start the server, invoke the <see cref="StartAsync"/> event;
+        /// </summary>
+        public Task OnLoadedAsync()
+        {
+            var token = _asyncListener.BeginAsyncOperation("OnLoadedAsync");
+
+            // set up event stream so that we start LSP server once Roslyn is loaded
+            _eventListener.WorkspaceStarted.ContinueWith(async _ =>
+            {
+                EnsureRemoteHost();
+
+                // wait until remote host is available before let platform know that they can activate our LSP
+                var client = await _workspace.TryGetRemoteHostClientAsync(CancellationToken.None).ConfigureAwait(false);
+                if (client == null)
+                {
+                    // there is no OOP. either user turned it off, or process got killed.
+                    // don't ask platform to start LSP
+                    return;
+                }
+
+                // let platform know that they can start us
+                await StartAsync.InvokeAsync(this, EventArgs.Empty).ConfigureAwait(false);
+            }, TaskScheduler.Default).CompletesAsyncOperation(token);
+
+            return Task.CompletedTask;
+
+            void EnsureRemoteHost()
```

Comment:
>EnsureRemoteHost [](start = 17, length = 16)

Inline - it's called just once.

Reply:
sure.. but many local function called only once though.

Reply:
I don't see much benefit of such functions unless they encapsulate some more complex logic.